clang/lib/CIR/Dialect/IR/CIRDialect.cpp

```diff
@@ -250,6 +251,29 @@ void AllocaOp::build(::mlir::OpBuilder &odsBuilder,
   odsState.addTypes(addr);
 }
 
+LogicalResult AllocaOp::verify() {
+  auto modOp = getOperation()->getParentOfType<ModuleOp>();
+  // If no data layout info specified, just skip it
+  if (!mlir::dyn_cast_or_null<mlir::DataLayoutSpecAttr>(
```

Comment:
Do we have other parts of the code using this pattern of getting the DL?
I wonder if we can create some unified interface implementation since we don’t need every new contributor to find the best method to get the DL for inspection :)
I’m not sure where it should be defined (in the dialect? In this file as static method? In another file?) but it for sure can help future us! 😇

Reply:
I believe you are overcomplicating things a bit, a `mlir::DataLayout layout(op->getParentOfType<mlir::ModuleOp>());` should suffice here. Datalayout is assumed to be available.

Reply:
I found that DataLayout is not available in the test cases from `test/CIR/IR`.

Reply:
@sitio-couto is going to introduce (a) a triple thing for CIR to compute datalayout and (b) the bits for it to work in some cases in cir-opt (i.e. when AST isn't available). If you need to test something ABI related you should check it coming from the source, not from standalone CIR.

Reply:
Yes, this is what I expected before. But to get current tests not failed in this verification, we have to skip the situations of missing DataLayout. What do you think?

Reply:
Sorry, I wasn't keeping a close look at the logic just yet. The verifier shouldn't be looking at datalayout here, the verifier only care the invariants of the operation make sense, if you are querying it from somewhere you are baking the wrong approach.

In fact I'm not really sure we need this PR at all - what are you trying to achieve here? Please clarify. If are you looking into `cir.alloca` having an optional address space attribute, it should be something akin to  `ConfinedAttr<OptionalAttr<I64Attr>, [IntMinValue<0>]>:$alignment,`, and the semantic during LLVM lowering could be similar: you either use what's specified in the operation during lowering or you query the datalayout. You could still have a helper as part of the extra declarations, but that should have nothing to do with verification. Does that make sense? Happy to try to clarify more too.

Reply:
* `AllocaOp` does not need an attribute of addspace, because we pass the result pointer type to its constructor, which already includes one.
* In clang, `CreateAlloca` seems to never specify individual addrspace. Instead, the `addrspace` is fetched from the target data layout.
* We could not mimic such approach immediately, because there's no actual need for "non-default (non-0) alloca address space" in CIR. (SPIR-V target does not need it.) I don't think it's a good idea to add this feature now.
* But to better track this missing feature, and also better track the tricky issue about `getASTAllocaAddressSpace()` in PR of SPIR-V target, I want to introduce this verification at early stage of CIR. It makes sure that we can find the root cause immediately after emitting incorrect AllocaOps.
* LLVM IR once introduced the same check in [this commit](https://github.com/llvm/llvm-project/commit/3c1fc768ed6cf2b01463df036ae6ae3b1f0de632), but removed now. We may also follow such a pace in the future.

It's closely associated with SPIR-V target. More details in [this comment](https://github.com/llvm/clangir/pull/671#discussion_r1632251899).

That's why it's a verifier, rather than an enhancement in CIRBuilder. Hope my expression is clear enough😃.

But I think you're right. `Op::verify` is for the operation itself, just like we do not use `Op::fold` to do match-and-rewrite. Do we have alternative mechanisms in MLIR?

Reply:
Interesting, so CIR's `AllocaOp` always carries the address space through its return type, in contrast to the LLVM `alloca` which has this [rule](https://llvm.org/docs/LangRef.html#id2056):
> If the address space is not explicitly specified, the object is allocated in the alloca address space from the [datalayout string](https://llvm.org/docs/LangRef.html#langref-datalayout).

Couldn't we just propagate the address space that clang determines (or is given in a lit test) through CIR? At LLVM emission time, we could check if the address space matches the LLVM module's default, and if not, set it explicitly on the alloca instruction.

Reply:
What is the missing feature you want to protect against? Do we need something like "default address space number is not 0" or so?

Reply:
Alloca is generated for local(automatic) variables, or compiler builtin function. The AST information from the frontend is VarDecl `int x;`, which does not carries pointers unfortunately.

Yes, currently `buildAlloca` always returns pointers with addrspace(0). But it's not blocking SPIR-V target if I'm not mistaken.

Reply:
> Interesting, so CIR's `AllocaOp` always carries the address space through its return type, in contrast to the LLVM `alloca`...

This is what I'm trying to suggest as well.

> Couldn't we just propagate the address space that clang determines (or is given in a lit test) through CIR?

Yep, exactly my point.

> At LLVM emission time, we could check if the address space matches the LLVM module's default, and if not, set it explicitly on the alloca instruction.

I'd say that it should behave like alignment: override if specified, otherwise use DL.

> What is the missing feature you want to protect against? Do we need something like "default address space number is not 0" or so?

Good point.

Seems like there's a lot of conversation around numbers regarding address space here and in the other PR. MLIR provides us enough resources to design decent abstractions to cope with silly restrictions, and also good to rememeber that CIR doesn't need to do exactly like LLVM does in number assumptions, we are free to be expressive in our representations (enum kinds, extra declarations, etc).